ocular/src/chain/client/cache.rs

```diff
@@ -0,0 +1,528 @@
+                    Err(err) => {
+                        return Err(CacheError::Toml(err.to_string()));
+                    }
+                };
+
+                dbg!(&toml);
+
+                for endpt in &toml.endpoint {
+                    endpoints.insert(endpt.address.to_string());
+                }
+            }
+        }
+
+        // Finally we can manipulate the actual file after checking the override settings
+        if override_if_exists || !std::path::Path::new(&path).exists() {
+            // Note this creates a new file or truncates the existing one
+            let _res = match File::create(&path) {
+                Ok(res) => res,
+                Err(err) => return Err(CacheError::FileIO(err.to_string())),
+            };
+        }
+
+        // If patch specified create
+        Ok(Cache {
+            grpc_endpoint_cache: Box::new(FileCache { path, endpoints }),
+        })
+    }
+
+    /// Constructor for in memory cache.
+    pub fn create_memory_cache(endpoints: Option<HashSet<String>>) -> Result<Cache, CacheError> {
+        let cache = match endpoints {
+            Some(endpoints) => MemoryCache { endpoints },
+            None => MemoryCache {
+                endpoints: HashSet::new(),
+            },
+        };
+
+        Ok(Cache {
+            grpc_endpoint_cache: Box::new(cache),
+        })
+    }
+}
+
+/// File based cache
+pub struct FileCache {
+    path: String,
+    endpoints: HashSet<String>,
+}
+
+impl GrpcCache for FileCache {
+    fn is_initialized(&self) -> bool {
+        !self.path.is_empty()
+    }
+
+    fn add_item(&mut self, item: String) -> Result<(), CacheError> {
+        if !self.endpoints.insert(item.clone()) {
+            return Err(CacheError::AlreadyExists(item));
+        }
+
+        let content = match std::fs::read_to_string(&self.path) {
+            Ok(result) => result,
+            Err(err) => {
+                return Err(CacheError::FileIO(err.to_string()));
+            }
+        };
+
+        let mut toml: GrpcEndpointToml = GrpcEndpointToml::default();
+
+        // Possible contents is empty, check to avoid parsing errors
+        if !content.is_empty() {
+            toml = match toml::from_str(&content) {
+                Ok(result) => result,
+                Err(err) => {
+                    return Err(CacheError::Toml(err.to_string()));
+                }
+            };
+
+            dbg!(&toml);
+        }
+
+        // Add new item
+        toml.endpoint.push(GrpcEndpoint { address: &item });
+
+        let toml_string = toml::to_string(&toml).expect("Could not encode toml value.");
+
+        dbg!(&toml_string);
+
+        // Rewrite file
+        std::fs::write(&self.path, toml_string).expect("Could not write to file.");
+
+        Ok(())
+    }
+
+    fn remove_item(&mut self, item: String) -> Result<(), CacheError> {
+        if !self.endpoints.remove(&item) {
+            return Err(CacheError::DoesNotExist(item));
+        }
```

Comment:
Yeah I think since it's a cache we don't care about the `bool` returned by the `HashSet` for `remove()`. I don't think the user will care that it didn't exist already.